packages/synchronize/lib/src/sync.dart

```diff
@@ -0,0 +1,263 @@
+  objectsB = await sources.sourceB.listObjects();
+  final entries = journal.entries.toList();
+  for (final entry in entries) {
+    final objectA = objectsA.find(entry.id);
+    final objectB = objectsB.find(entry.id);
+
+    // Remove all entries from journal that don't exist anymore
+    if (objectA == null && objectB == null) {
+      journal.entries.removeWhere((final e) => e.id == entry.id);
+      continue;
+    }
+
+    if (objectA != null && objectB != null) {
+      final changedA = entry.etagA != await sources.sourceA.getObjectETag(objectA);
+      final changedB = entry.etagB != await sources.sourceB.getObjectETag(objectB);
+
+      if (changedA && changedB) {
+        conflicts.add(
+          SyncConflict<T1, T2>(
+            id: objectA.id,
+            type: SyncConflictType.bothChanged,
+            objectA: objectA,
+            objectB: objectB,
+          ),
+        );
+        continue;
+      }
+
+      if (changedA && !changedB) {
+        actions.add(SyncActionWriteToB<T1, T2>(objectA));
+        continue;
+      }
+
+      if (changedB && !changedA) {
+        actions.add(SyncActionWriteToA<T1, T2>(objectB));
+        continue;
+      }
+    }
+  }
+
+  final unsolvedConflicts = <SyncConflict<T1, T2>>[];
+  for (final conflict in conflicts) {
+    final solution = conflictSolutions[conflict.id] ?? sources.findSolution(conflict.objectA, conflict.objectB);
+    if (solution != null) {
+      switch (solution) {
+        case SyncConflictSolution.overwriteA:
+          actions.add(SyncActionWriteToA<T1, T2>(conflict.objectB));
+          break;
+        case SyncConflictSolution.overwriteB:
+          actions.add(SyncActionWriteToB<T1, T2>(conflict.objectA));
+          break;
+        case SyncConflictSolution.skip:
+          if (keepSkipsAsConflicts) {
+            unsolvedConflicts.add(
+              SyncConflict<T1, T2>(
+                id: conflict.id,
+                type: conflict.type,
+                objectA: conflict.objectA,
+                objectB: conflict.objectB,
+                skipped: true,
+              ),
+            );
+          }
+          break;
+      }
+    } else {
+      unsolvedConflicts.add(conflict);
+    }
+  }
+
+  return SyncDiff<T1, T2>(
+    _sortActions(actions),
+    unsolvedConflicts,
+  );
+}
+
+List<SyncAction<T1, T2>> _sortActions<T1, T2>(final List<SyncAction<T1, T2>> actions) {
+  final addActions = <SyncAction<T1, T2>>[];
+  final removeActions = <SyncAction<T1, T2>>[];
+  for (final action in actions) {
+    if (action is SyncActionWriteToA<T1, T2>) {
+      addActions.add(action);
+    } else if (action is SyncActionWriteToB<T1, T2>) {
+      addActions.add(action);
+    } else if (action is SyncActionDeleteFromA<T1, T2>) {
+      removeActions.add(action);
+    } else if (action is SyncActionDeleteFromB<T1, T2>) {
+      removeActions.add(action);
+    } else {
+      throw Exception('illegal action for sorting');
+    }
+  }
```

Comment:
Once SyncAction is sealed you should use a switch statement. They are guaranteed to be exhaustive.